### PR TITLE
#4260 - Add ability to turn off ScheduleTypeLimits enforcement when already assigned

### LINKED ISSUE #4260: space load objects only allow fractional schedules
Issue overview
--------------
EnergyPlus allows non-fractional schedules for Lights and ElectricEquipment objects (and maybe others) using total power, but OpenStudio does not. 

For example, I want to have Lights object with power = 1 W and apply a ScheduleFile of power measurements to it, but OpenStudio does not set the schedule because the ScheduleFile has a non-fractional ScheduleTypeLimits and EnergyPlus fails due to a missing schedule. The reverse, power = peak power with schedule of fractional power, runs fine. In the former case, setting the schedule in the IDF runs fine without errors or warnings.

```
test = false

require 'openstudio'
include OpenStudio::Model

m = Model.new

stl = ScheduleTypeLimits.new(m)
if test
  stl.setLowerLimitValue(0)
  stl.setUpperLimitValue(1)
end

sr = ScheduleRuleset.new(m)
sr.setScheduleTypeLimits(stl)

ld = LightsDefinition.new(m)
ld.setLightingLevel(1)

l = Lights.new(ld)
l.setSchedule(sr)

puts l
```

## Current Behavior
Set `test = false` above.

```
OS:Lights,
  {aa30d18c-5c71-4336-83cf-5bdbefab6f22}, !- Handle
  Lights 1,                               !- Name
  {58ea5d10-b57d-4316-bad4-ad1cdbd77491}, !- Lights Definition Name
  ,                                       !- Space or SpaceType Name
  ,                                       !- Schedule Name
  1,                                      !- Fraction Replaceable
  ,                                       !- Multiplier
  General;                                !- End-Use Subcategory
```

## Expected Behavior

Set `test = true` above.

```
OS:Lights,
  {02547a43-215c-4114-a5d1-531727fabb9a}, !- Handle
  Lights 1,                               !- Name
  {4de8a137-9dc8-48f6-b80a-404d0f1729cf}, !- Lights Definition Name
  ,                                       !- Space or SpaceType Name
  {b2542f69-155a-4262-ab19-587eea1b98c2}, !- Schedule Name
  1,                                      !- Fraction Replaceable
  ,                                       !- Multiplier
  General;                                !- End-Use Subcategory
```

## Steps to Reproduce

1.
2.
3.
4.

### Environment

Some additional details about your environment for this issue (if relevant):
 - Platform (Operating system, version):
 - Version of OpenStudio (if using an intermediate build, include SHA): 3.1.0

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

 - Fixes #4260 - Add ability to turn off ScheduleTypeLimits enforcement when already assigned

### Pull Request Author

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Checked behavior in OpenStudioApplication, adjusted policies as needed (`src/openstudio_lib/library/OpenStudioPolicy.xml`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: src/model/ScheduleTypeRegistry.hpp [hunk 1]
```diff
@@ -108,12 +108,19 @@ namespace model {
 
     std::string getDefaultName(const ScheduleType& scheduleType) const;
 
+    bool enforceScheduleTypeLimits() const;
+
+    // This is true by default. Set to false to avoid checking already-assigned ScheduleTypeLimits
+    // If you do not have a ScheduleTypeLimits assigned, one will still be assigned for you
+    bool setEnforceScheduleTypeLimits(bool enforceScheduleTypeLimits);
```
Comment (jmarrec) [inline src/model/ScheduleTypeRegistry.hpp:115]: Add a new getter and setter.

File: src/model/ScheduleTypeRegistry.hpp [hunk 2]
```diff
@@ -108,12 +108,19 @@ namespace model {
...
+
    private:
     REGISTER_LOGGER("openstudio.model.ScheduleTypeRegistry");
     ScheduleTypeRegistrySingleton();
 
     typedef std::map<std::string, std::vector<ScheduleType>> ClassNameToScheduleTypesMap;
     ClassNameToScheduleTypesMap m_classNameToScheduleTypesMap;
+    bool m_enforceScheduleTypeLimits = true;
```
Comment (jmarrec) [inline src/model/ScheduleTypeRegistry.hpp:123]: corresponding private member, defaults to true, business as usual

File: src/model/ScheduleTypeRegistry.cpp [hunk 3]
```diff
@@ -619,7 +628,14 @@ namespace model {
...
       if (!isCompatible(scheduleType, *scheduleTypeLimits, false)) {
-        result = false;
+        // Skip if user has disabled it
+        if (ScheduleTypeRegistry::instance().enforceScheduleTypeLimits()) {
+          result = false;
+        } else {
+          LOG_FREE(Warn, "checkOrAssignScheduleTypeLimits",
+                   "Schedule '" << schedule.nameString() << "' has a ScheduleTypeLimits named '" << scheduleTypeLimits->nameString()
+                                << "' that is not compatible, but you have turned off enforcement.");
+        }
```
Comment (jmarrec) [inline src/model/ScheduleTypeRegistry.cpp:638]: If enforce turned off:
* If provided ScheduleTypeLimits is not compatible, still assign but warn.
* If no ScheduleTypeLimits provided, still assign one

File: src/model/ModelResources.i [hunk 4]
```diff
@@ -265,6 +265,7 @@ SWIG_MODELOBJECT(CoilCoolingDXCurveFitOperatingMode, 1);
 SWIG_MODELOBJECT(CoilCoolingDXCurveFitSpeed, 1);
 
 %include <model/ScheduleTypeRegistry.hpp>
+%template(ScheduleTypeRegistry) openstudio::Singleton<openstudio::model::ScheduleTypeRegistrySingleton>;
```
Comment (jmarrec) [inline src/model/ModelResources.i:268]: Swig the ScheduleTypeRegistry Singleton. In ruby do `OpenStudio::Model::ScheduleTypeRegistry::instance.setEnforceScheduleTypeLimits(false)`

**Comment by tijcolem** [comment]
@jmarrec Looks good. Windows did pick up another failure unit test. 

`2063 - ModelFixture.ScheduleTypeRegistry_TurnOffEnforce (Failed)`

Linux and Mac seem happy.

**Comment by jmarrec** [comment]
@tijcolem Thanks for letting me know about windows... I had wrongly assumed the DateTest was the one failing (we should disable it on windows... we need 100% green to be able to enforce clean CI results).

I'll carve some time out to go see why windows is trying to ruin my life again in the coming days.

**Comment by jmarrec** [comment]
Ruby: `l.setSchedule(sr)`
ScheduleTypeRegistrySingleton address is this = 0x00007ffd9816a238

Ruby: `ScheduleTypeRegistry::enforceScheduleTypeLimits`
ScheduleTypeRegistrySingleton address is this = 0x00007ffda4edad28


0x00007ffd9816a238 != 0x00007ffda4edad28. What a singleton!

**Comment by jmarrec** [comment]
After deliberation, this feature is being dropped. Exposing the Singleton through SWIG is a bad idea that will come to bite us, and is not worth the time nor hassle to support a few corner cases that can be worked around by using setPointer directly.

**Pull request closed without merging**